File: Tree/BinarySearchTree/BinarySearchTree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self._insert_recursive(self.root, value)

    def _insert_recursive(self, current, value):
        if current is None:
            return Node(value)
        if value < current.value:
            current.left = self._insert_recursive(current.left, value)
        elif value > current.value:
            current.right = self._insert_recursive(current.right, value)
        return current

    def search(self, value):
        return self._search_recursive(self.root, value)

    def _search_recursive(self, current, value):
        if current is None:
            return False
        if current.value == value:
            return True
        if value < current.value:
            return self._search_recursive(current.left, value)
        return self._search_recursive(current.right, value)

    def find_min(self, node=None):
        current = node if node else self.root
        if current is None:
            return None
        while current.left is not None:
            current = current.left
        return current.value

    def find_max(self, node=None):
        current = node if node else self.root
        if current is None:
            return None
        while current.right is not None:
            current = current.right
        return current.value

    def delete(self, value):
        self.root = self._delete_recursive(self.root, value)

    def _delete_recursive(self, current, value):
        if current is None:
            return current

        if value < current.value:
            current.left = self._delete_recursive(current.left, value)
        elif value > current.value:
            current.right = self._delete_recursive(current.right, value)
        else:
            # Case 1 & 2: Node has one child or no children
            if current.left is None:
                return current.right
            elif current.right is None:
                return current.left

            # Case 3: Node has two children
            # Find the in-order successor (smallest in the right subtree)
            successor_value = self.find_min(current.right)
            current.value = successor_value
            # Delete the duplicate successor node from the right branch
            current.right = self._delete_recursive(current.right, successor_value)

        return current

    def get_height(self, node=None):
        if node is None:
            return -1  # Height of empty tree is -1
        left_height = self.get_height(node.left)
        right_height = self.get_height(node.right)
        return max(left_height, right_height) + 1

    def traverse_in_order(self, node):
        if node:
            self.traverse_in_order(node.left)
            print(node.value, end=" ")
            self.traverse_in_order(node.right)

if __name__ == "__main__":
    bst = BinarySearchTree()
    nodes = [50, 30, 70, 20, 40, 60, 80]
    for val in nodes:
        bst.insert(val)

    print("Initial In-Order traversal (Sorted):")
    bst.traverse_in_order(bst.root)  # 20 30 40 50 60 70 80
    print(f"\nMinimum: {bst.find_min()} | Maximum: {bst.find_max()}")
    print(f"Tree Height: {bst.get_height(bst.root)}")

    print("\n--- Deleting node 30 (Has two children: 20 and 40) ---")
    bst.delete(30)
    bst.traverse_in_order(bst.root)  # 20 40 50 60 70 80
    print()

File: Tree/BinarySearchTree/BinarySearchTree.java
public class BinarySearchTree {
    private static class Node {
        int value;
        Node left, right;
        Node(int value) {
            this.value = value;
            this.left = this.right = null;
        }
    }

    private Node root = null;

    public void insert(int value) {
        root = insertRecursive(root, value);
    }

    private Node insertRecursive(Node current, int value) {
        if (current == null) return new Node(value);
        if (value < current.value) current.left = insertRecursive(current.left, value);
        else if (value > current.value) current.right = insertRecursive(current.right, value);
        return current;
    }

    public int findMin() {
        if (root == null) throw new IllegalStateException("Tree is empty");
        return findMinNode(root).value;
    }

    private Node findMinNode(Node node) {
        Node current = node;
        while (current.left != null) current = current.left;
        return current;
    }

    public void delete(int value) {
        root = deleteRecursive(root, value);
    }

    private Node deleteRecursive(Node current, int value) {
        if (current == null) return null;

        if (value < current.value) {
            current.left = deleteRecursive(current.left, value);
        } else if (value > current.value) {
            current.right = deleteRecursive(current.right, value);
        } else {
            // Case 1 & 2
            if (current.left == null) return current.right;
            if (current.right == null) return current.left;

            // Case 3
            Node successor = findMinNode(current.right);
            current.value = successor.value;
            current.right = deleteRecursive(current.right, successor.value);
        }
        return current;
    }

    public int getHeight() {
        return getHeightRecursive(root);
    }

    private int getHeightRecursive(Node node) {
        if (node == null) return -1;
        return Math.max(getHeightRecursive(node.left), getHeightRecursive(node.right)) + 1;
    }

    public void printInOrder() {
        inOrder(root);
        System.out.println();
    }

    private void inOrder(Node node) {
        if (node != null) {
            inOrder(node.left);
            System.out.print(node.value + " ");
            inOrder(node.right);
        }
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();
        int[] data = {50, 30, 70, 20, 40, 60, 80};
        for (int x : data) bst.insert(x);

        System.out.print("Tree In-Order: ");
        bst.printInOrder();
        System.out.println("Tree Height: " + bst.getHeight());

        System.out.println("Deleting node 20 (Leaf)...");
        bst.delete(20);
        bst.printInOrder();
    }
}
